tree/ntuple/v7/src/RPageStorageFile.cxx

```diff
@@ -125,48 +125,107 @@ ROOT::Experimental::Internal::RPageSinkFile::CommitSealedPageImpl(DescriptorId_t
          // Skip empty ranges, they might not have a physical column ID!
          continue;
       }
 
       const auto bitsOnStorage = RColumnElementBase::GetBitsOnStorage(
          fDescriptorBuilder.GetDescriptor().GetColumnDescriptor(range.fPhysicalColumnId).GetModel().GetType());
-      for (auto sealedPageIt = range.fFirst; sealedPageIt != range.fLast; ++sealedPageIt) {
-         size += sealedPageIt->GetBufferSize();
-         bytesPacked += (bitsOnStorage * sealedPageIt->GetNElements() + 7) / 8;
-      }
-   }
-   if (bytesPacked >= fOptions->GetMaxKeySize()) {
-      // Cannot fit it into one key, fall back to one key per page.
-      return RPagePersistentSink::CommitSealedPageVImpl(ranges);
-   }
 
-   Detail::RNTupleAtomicTimer timer(fCounters->fTimeWallWrite, fCounters->fTimeCpuWrite);
-   // Reserve a blob that is large enough to hold all pages.
-   std::uint64_t offset = fWriter->ReserveBlob(size, bytesPacked);
-
-   // Now write the individual pages and record their locators.
-   std::vector<ROOT::Experimental::RNTupleLocator> locators;
-   for (auto &range : ranges) {
       for (auto sealedPageIt = range.fFirst; sealedPageIt != range.fLast; ++sealedPageIt) {
-         fWriter->WriteIntoReservedBlob(sealedPageIt->GetBuffer(), sealedPageIt->GetBufferSize(), offset);
-         RNTupleLocator locator;
-         locator.fPosition = offset;
-         locator.fBytesOnStorage = sealedPageIt->GetDataSize();
-         locators.push_back(locator);
-         offset += sealedPageIt->GetBufferSize();
+         if (batch.fSize > 0 && batch.fSize + sealedPageIt->GetBufferSize() > maxKeySize) {
+            /**
+             * Adding this page would exceed maxKeySize. Since we always want to write into a single key
+             * with vectorized writes, we commit the current set of pages before proceeding.
+             * NOTE: we do this *before* checking if sealedPageIt->GetBufferSize() > maxKeySize to guarantee that
+             * we always flush the current batch before doing an individual WriteBlob. This way we
+             * preserve the assumption that a CommittedBatch always contain a sequential set of pages.
+             */
+            batch.fEnd = std::make_pair(std::next(rangeIt), sealedPageIt);
+            CommitBatchOfPages(batch, locators);
+         }
+
+         if (sealedPageIt->GetBufferSize() > maxKeySize) {
+            // This page alone is bigger than maxKeySize: save it by itself, since it will need to be
+            // split into multiple keys.
+
+            // Since this check implies the previous check on batchSize + newSize > maxSize, we should
+            // already have committed the current batch before writing this page.
+            assert(batch.fSize == 0);
+
+            std::uint64_t offset = fWriter->WriteBlob(sealedPageIt->GetBuffer(), sealedPageIt->GetBufferSize(),
+                                                      (bitsOnStorage * sealedPageIt->GetNElements() + 7) / 8);
```

Comment:
We can factor out the expression `(bitsOnStorage * sealedPageIt->GetNElements() + 7) / 8)` into a named const.